driver-core/src/main/com/mongodb/internal/async/client/ClientSessionBinding.java

```diff
@@ -58,19 +59,27 @@ public ReadPreference getReadPreference() {
 
     @Override
     public void getReadConnectionSource(final SingleResultCallback<AsyncConnectionSource> callback) {
-        if (isConnectionSourcePinningRequired()) {
-            getPinnedConnectionSource(true, callback);
-        } else {
-            wrapped.getReadConnectionSource(new WrappingCallback(callback));
-        }
+        isConnectionSourcePinningRequired((isConnectionSourcePinningRequired, t) -> {
+            if (t != null) {
+                callback.onResult(null, t);
+            } else if (isConnectionSourcePinningRequired) {
+                getPinnedConnectionSource(true, callback);
+            } else {
+                wrapped.getReadConnectionSource(new WrappingCallback(callback));
+            }
+        });
```

Comment:
I had a hard time parsing this code, maybe because of the method name and variable name being the same, combined with the extra layer of callbacks.  What do you think about just wrapping it like this?

```suggestion

        try {
            if (isConnectionSourcePinningRequired()) {
                getPinnedConnectionSource(true, callback);
            } else {
                wrapped.getReadConnectionSource(new WrappingCallback(callback));
            }
        } catch (Exception e) {
            callback.onResult(null, e);
        }
```

Reply:
I tried that and also tried abstracting the pattern  eg:

```
import java.util.function.Consumer;

public final class CallbackHelper {

    public static <T> void guardCallback(
            final Consumer<SingleResultCallback<T>> callbackConsumer,
            final SingleResultCallback<T> callback) {
        try {
            callbackConsumer.accept(callback);
        } catch (Throwable t) {
            callback.onResult(null, t);
        }
    }

    private CallbackHelper() {
    }
}
```

However, I discounted it as I wanted to be explicit that `isConnectionSourcePinningRequired()` can produce an error.  Also I didnt want to wrap the same method call in try catches twice.

Reply:
ok, i can live with it.